### PR TITLE
Issue 414/add message to contacts

### LINKED ISSUE #414: Enhancement: Add "Message" action to contacts list
Currently, the contacts list only has a few actions: go to pod, pin contact, and delete

image

Another useful action to have on this table would be to message the client. Add a button to each contact in this list that launches  the "send message" modal when clicked, that prepopulates the "to" field with the webId of the contact.

### LINKED ISSUE #422: Incorrect jsDoc comments in `AddContactModal.jsx`
### Bug Title:
Fix incorrect jsDoc in `AddContactModal.jsx`

---
### Bug Description:
The jsDoc for `renderWebId` function at top of file is inaccurate.
The jsDoc for `AddContactModal` is inaccurate.

### PR DESCRIPTION
**Pull request by andycwilliams**
## This PR:
Resolves #414 

I've been focused elsewhere and am only now getting back into the swing of PASS, so please make sure this is proper.

**1.** Adds a message button to each row in the contacts list
**2.** This button auto populates the `To` field with the corresponding username

## Screenshots (if applicable):

Capture
Capture2

## Issues needing discussion/feedback (optional):
**1.** Determine whether this code could be more efficient
**2.** Determine if the message icon is adequate or should be replaced

### CODE REVIEW

**Comment by leekahung** [comment]
At the moment, the send message function hard-codes the trailing `profile/card#me` that comes with all solid webIds so you only need to enter the pod URL for the `To` field, sort of like not needing to type `www.` for websites.

Was thinking of modifying the UI for the `To` field to make that more apparent (like with a placeholder or text at the end of the field).

File: src/components/Contacts/ContactListTable.jsx [hunk 1]
```diff
@@ -62,6 +72,15 @@ const ContactListTable = ({ contacts, deleteContact }) => {
       headerAlign: 'center',
       align: 'center'
     },
+    {
+      field: 'Message',
+      renderCell: (contactId) => <SendIcon onClick={() => handleSendMessage(contactId)} />,
+      sortable: false,
+      filterable: false,
+      width: 80,
```
Comment (leekahung) [inline src/components/Contacts/ContactListTable.jsx:85]: Think we could have all the columns with interactive buttons be the same width. So perhaps we could make the other two columns 80px width as well.

**Comment by andycwilliams** [comment]
> At the moment, the send message function hard-codes the trailing `profile/card#me` that comes with all solid webIds so you only need to enter the pod URL for the `To` field, sort of like not needing to type `www.` for websites.
> 
> Was thinking of modifying the UI for the `To` field to make that more apparent (like with a placeholder or text at the end of the field).

Do you mean you'd like to do that in this PR, or just that it's something planned for the future?

> Think we could have all the columns with interactive buttons be the same width. So perhaps we could make the other two columns 80px width as well.

Done in next push.

**Comment by leekahung** [comment]
> > At the moment, the send message function hard-codes the trailing `profile/card#me` that comes with all solid webIds so you only need to enter the pod URL for the `To` field, sort of like not needing to type `www.` for websites.
> > Was thinking of modifying the UI for the `To` field to make that more apparent (like with a placeholder or text at the end of the field).
> 
> Do you mean you'd like to do that in this PR, or just that it's something planned for the future?

Think we can do it in a later PR.

**Comment by andycwilliams** [comment]
> Think we can do it in a later PR.

Sounds good. Is there anything more needed for this?

**Comment by leekahung** [comment]
> > Think we can do it in a later PR.
> 
> Sounds good. Is there anything more needed for this?

If it's still using webId for the "To" field, it needs to be updated to use Pod URL (unless we update the sendMessage function in utils to just accept straight up webIds as well). Other wise think it should be ready?

Is there also a way to make the send icon gray like the other icons? Makes them consistent to the other icons on the row.

**Comment by andycwilliams** [comment]
> If it's still using webId for the "To" field, it needs to be updated to use Pod URL (unless we update the sendMessage function in utils to just accept straight up webIds as well). Other wise think it should be ready?

So just removing `profile/card#me` from `https://exampleusername.solidcommunity.net/profile/card#me`? That should be simple

> Is there also a way to make the send icon gray like the other icons? Makes them consistent to the other icons on the row.

Yep, that's easy.

**Comment by leekahung** [comment]
> > If it's still using webId for the "To" field, it needs to be updated to use Pod URL (unless we update the sendMessage function in utils to just accept straight up webIds as well). Other wise think it should be ready?
> 
> So just removing `profile/card#me` from `https://exampleusername.solidcommunity.net/profile/card#me`? That should be simple

Yeah, I think you could just pull contact.podUrl directly from the contact object:

Screen Shot 2023-10-31 at 19 47 53

**Comment by andycwilliams** [comment]
> Yeah, I think you could just pull contact.podUrl directly from the contact object:

Should be good to go now?

**Review by leekahung — CHANGES_REQUESTED** [review]
Had previously thought it was good to go, but it looks like there still a few kinks to work out.

The first being that I couldn't send any message and the error not entering a recipient pod URL appeared. Looking at `NewMessageModal.jsx`, it's likely that `sendMessageTTL` function in Line 83 did not trigger. The `handleSubmit` function needs to take into consideration the new `toField` variable as well.

Second, the cursor for the send icon remains `cursor` instead of `pointer`. This one should be easier to handle.

https://github.com/codeforpdx/PASS/assets/14917816/e8e1e69a-93c2-4e53-a4ad-5dbf2a6aee2a

**Comment by leekahung** [comment]
> Had previously thought it was good to go, but it looks like there still a few kinks to work out.
> 
> The first being that I couldn't send any message and the error not entering a recipient pod URL appeared. Looking at `NewMessageModal.jsx`, it's likely that `sendMessageTTL` function in Line 83 did not trigger. The `handleSubmit` function needs to take into consideration the new `toField` variable as well.
> 
> Second, the cursor for the send icon remains `cursor` instead of `pointer`. This one should be easier to handle.
> 
>  Screen.Recording.2023-11-02.at.0.20.36.mov

Thinking about it again, I think we could also disable the `To` field since we don't really need to change that if it's going to be a direct message to that person from the contact list.

File: src/components/Modals/NewMessageModal.jsx [hunk 2]
```diff
@@ -32,11 +32,12 @@ import { MessageContext, SignedInUserContext } from '@contexts';
  * @param {boolean} Props.showModal - Boolean for showing message modal
  * @param {React.Dispatch<React.SetStateAction<boolean>>} Props.setShowModal
  * - React set function for showModal
+ * @param Props.toField
```
Comment (leekahung) [inline src/components/Modals/NewMessageModal.jsx]: The JSDoc for Line 35 needs to be corrected. It's missing the type and a description. The following format should do:

```js
/*
 * @param {variable_type} variable_in_question - Description of variable
 */
```

You could check the other `@param` lines in the JSDoc for reference.

Comment (andycwilliams) [reply src/components/Modals/NewMessageModal.jsx]: All the changes should be implemented now sans the `sendMessageTTL`. Still working on that.

Also wanted to mention that going to the Messages tab gives this error:
```
MessagePreview.jsx:100 Warning: Each child in a list should have a unique "key" prop.
```

Comment (leekahung) [reply src/components/Modals/NewMessageModal.jsx]: > All the changes should be implemented now sans the `sendMessageTTL`. Still working on that.
> 
> Also wanted to mention that going to the Messages tab gives this error:
> 
> ```
> MessagePreview.jsx:100 Warning: Each child in a list should have a unique "key" prop.
> ```

This error is related to the rendering of the message list. @bingeboy got that fixed in PR #460. We can ignore that here.

**Review by leekahung — APPROVED** [review]
Just confirmed the changes. Functionality works now.

Also, not really a blocker, since Line 159 in `src/components/Modals/NewMessageModal.jsx` with the double `!` works logically, but a bit confusing to read. Think it'll be better to change it to `toField !== ""`.

**Pull request merged**